client/README.md

```diff
@@ -0,0 +1,1013 @@
+This information is structured in several merkle tree structures and a "ledger root" is signed by a quorum of validators.
+
+There are two types of merkle trees: **append-only accumulators** (TODO: more explanation) and **sparse merkle trees**. (TODO: jellyfish?)
+
+The ledger root authenticates an append-only tree of "transaction infos". 
+Each *transaction info* contains:
+
+* A signed transaction hash: this does not contain the transaction itself, but a hash of the signed transaction. (TODO: malleability warning?)
+* A state root: the root of a sparse merkle tree, where leafs are account states
+* A event root: the root of an append-only tree, where leafs are events (TODO: is this correct?)
+* gas used: gas used by the execution of the transaction
+* major status: this indicates if the transaction was excecuted successfuly
+    
+(TODO: diagram)
+
+To recapitulate, there are three trees:
+
+* A ledger info tree: append-only.
+* An event tree: append-only.
+* A state tree: sparse.
+
+Proofs for each type of tree can be inclusion or non-inclusion proofs.
+(TODO: what's a proof?)
+(TODO: explanation for each type of tree, what is the path? right/left nodes?)
+
+### Bitmaps
+
+A proof for a Merkle Tree includes information on the path a leaf or internal node takes to the root.
+A way to encode this path is to use a string of bits, where every bit indicates if the next node is on the left or on the right.
+Using a bitstring as a datastructure to compress our information into bits is called a **bitmap**.
+Libra defines bitmaps this way:
+
+* The most-left bits hold information about nodes near the root.
+* The most-right bits hold information about nodes near the leaf.
+
+In addition, to verify a proof, a client needs to know the siblings of each step of the path up to the root.
+These siblings are sometimes legitimate nodes, or placeholders. 
+As such, proofs come with a bitmap that indicates if the next sibling is a placeholder or not.
+
+<!-- diagram of sibling + bitmap in a merkle tree -->
+
+First, we define `node_type` which indicates if a sibling node in the tree is a placeholder (default) or not. It also indicates if the end of the bitmap has been reached.
+
+```c
+Enum node_type {
+    default     // if bit is 0
+    non_default // if bit is 1
+    end         // end of the bitmap
+}
+```
+
+Second, we define `node_position` which indicates if a node is on the left or on the right.
+
+```c
+Enum node_position {
+    left    // if bit is 0
+    right   // if bit is 1
+    end     // end of the bitmap
+}
+```
+
+There are two types of bitmaps used by Libra:
+
+* Bitmaps of 8 bytes.
+* Bitmaps of 32 bytes.
+
+We abstract each type of bitmap in this specification via `Bitmap8` and `Bitmap32`:
+
+**`Bitmap8(bitmap: uint64)`** takes a bitmap of 8 bytes. A function `Next()` acts on this abstraction to provide a `node_type`, reading bits from the right. Leading zeros are ignored.
+
+For example:
+
+```php
```

Comment:
If we have examples we should probably have the examples be in Rust and not php

Reply:
(that's not php but it made the syntax highlightning work better for some reason :D)

also, you cannot test this example as this is an abstraction

Reply:
You should write your pseudocode in a rust like syntax ;P

Reply:
that's what I'm doing with the consensus specification at the moment btw, some thoughts:

* it's easier to maintain, especially as the code is moving fast
* I find it not great for explicitly talking about errors that need to be handled (although I can add `ensure!()` and `return Err()` ...)
* it's also not great for people who don't know Rust
* not sure what to do at this point, but I'm writing the spec with both style atm just to see what sticks